### PR TITLE
Update splash screen logic

### PR DESCRIPTION
**Pull request by zoontek**
Following https://github.com/Expensify/App/issues/5620

### Details

There are several issue with the current splash integrations:

- Multiple `UIViewController` could be mounted at the same time (splash screen & growl notifications, could cause https://github.com/zoontek/react-native-bootsplash/issues/119)
- `bootsplash.hide()` could reject, which is not handled (see https://github.com/zoontek/react-native-bootsplash/issues/264, which happen on tapping notification on iOS 15+)
- It doesn't wait for `NavigationContainer` `onReady` callback to hide.
- It will not hide if something happen during context providers initialisation.

### Fixed Issues

https://github.com/Expensify/App/issues/5620

### QA Steps

- Starts the app on Android >= 5, check if the splash screen is visible
- Starts the app on iOS >= 10, check if the splash screen is visible
- Try popping an alert / a native modal once the screen is loaded
- Open the app using a push notification tap (on both iOS and Android)

### Tested On

- [x] Web
- [x] Mobile Web
- [x] Desktop
- [x] iOS
- [x] Android

#### Mobile Web

Screenshot 2021-12-01 at 10 30 00

### CODE REVIEW

File: android/app/build.gradle [hunk 1]
```diff
@@ -209,6 +209,7 @@ dependencies {
     implementation "com.facebook.react:react-native:+"  // From node_modules
 
     implementation "androidx.swiperefreshlayout:swiperefreshlayout:1.0.0"
+    implementation "androidx.core:core-splashscreen:1.0.0-alpha02"
```
Comment (zoontek) [inline android/app/build.gradle:212]: Needed to extends the `Theme.SplashScreen` theme in `styles.xml`

File: android/app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -31,12 +31,6 @@
         android:launchMode="singleTask"
         android:exported="true"
         android:windowSoftInputMode="adjustResize">
-      </activity>
-
-      <activity
```
Comment (zoontek) [inline android/app/src/main/AndroidManifest.xml:36]: Remove the need for a custom `Activity` since `react-native-bootsplash` v4 uses Android `SplashScreen` API.

File: android/app/src/main/res/values/colors.xml [hunk 3]
```diff
@@ -1,6 +1,7 @@
 <resources>
     <color name="bootsplash_background">#FFFFFF</color>
-    <color name="white">#FAFAFA</color>
+    <color name="white">#FFFFFF</color>
```
Comment (zoontek) [inline android/app/src/main/res/values/colors.xml:3]: Fix status bar being a different color than the splash screen background.

File: android/app/src/main/res/values/styles.xml [hunk 4]
```diff
@@ -7,11 +7,15 @@
         <item name="android:colorEdgeEffect">@color/gray4</item>
         <item name="colorAccent">@color/accent</item>
         <item name="android:statusBarColor">@color/white</item>
-        <item name="android:windowLightStatusBar">true</item>
+        <item name="android:windowLightStatusBar" tools:ignore="NewApi">true</item>
```
Comment (zoontek) [inline android/app/src/main/res/values/styles.xml:10]: `android:windowLightStatusBar` and `android:windowLightNavigationBar` are not available on Android 21 (the min version), `tools:ignore` silents the Android studio linter.

File: android/build.gradle [hunk 5]
```diff
@@ -4,8 +4,8 @@ buildscript {
     ext {
         buildToolsVersion = "29.0.2"
         minSdkVersion = 21
-        compileSdkVersion = 30
-        targetSdkVersion = 30
+        compileSdkVersion = 31
```
Comment (zoontek) [inline android/build.gradle:7]: Required for Android 12 support

File: ios/NewExpensify/AppDelegate.m [hunk 6]
```diff
@@ -54,12 +54,12 @@ - (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(
   self.window.rootViewController = rootViewController;
   [self.window makeKeyAndVisible];
 
+  [RNBootSplash initWithStoryboard:@"BootSplash" rootView:rootView]; // <- initialization using the storyboard file name
```
Comment (zoontek) [inline ios/NewExpensify/AppDelegate.m:57]: The splash screen should be init immediately after `RCTRootView` creation

File: src/App.js [hunk 7]
```diff
@@ -22,20 +22,21 @@ LogBox.ignoreLogs([
...
-            SafeArea,
-            LocaleContextProvider,
-            HTMLEngineProvider,
-        ]}
-    >
+    <ErrorBoundary errorMessage="NewExpensify crash caught by error boundary">
         <CustomStatusBar />
-        <ErrorBoundary errorMessage="NewExpensify crash caught by error boundary">
+
+        <ComposeProviders
```
Comment (zoontek) [inline src/App.js:28]: As an error could happen on context provider initialization, `ErrorBoundary` should wrap them to hide the splash screen is something goes wrong.

File: src/Expensify.js [hunk 8]
```diff
@@ -81,12 +82,8 @@ class Expensify extends PureComponent {
         // Run any Onyx schema migrations and then continue loading the main app
         migrateOnyx()
             .then(() => {
-                // When we don't have an authToken we'll want to show the sign in screen immediately so we'll hide our
-                // boot screen right away
-                if (!this.getAuthToken()) {
```
Comment (zoontek) [inline src/Expensify.js:86]: We wait for `react-navigation` to be ready. No need to hide the splash screen here

File: src/components/ErrorBoundary/BaseErrorBoundary.js [hunk 9]
```diff
@@ -2,18 +2,18 @@ import React from 'react';
 import PropTypes from 'prop-types';
 
 const propTypes = {
-    /* A message posted to `logError` (along with error data) when this component intercepts an error */
```
Comment (zoontek) [inline src/components/ErrorBoundary/BaseErrorBoundary.js:5]: Renamed because it's not only used for logging anymore.
Naming from https://github.com/bvaughn/react-error-boundary

File: src/components/ErrorBoundary/index.native.js [hunk 10]
```diff
@@ -1,16 +1,20 @@
...
     // Log the error to the server
     Log.alert(errorMessage, {error: error.message, errorInfo}, false);
 
     /* On native we also log the error to crashlytics
     * Since the error was handled we need to manually tell crashlytics about it */
     crashlytics().log(`errorInfo: ${JSON.stringify(errorInfo)}`);
     crashlytics().recordError(error);
+
+    // We hide the splash screen since the error might happened during app init
+    BootSplash.hide({fade: true});
```
Comment (zoontek) [inline src/components/ErrorBoundary/index.native.js:17]: Hide the splash screen in case something wrong happened (ideally, the `ErrorBoundary` should show an error screen with a CTA button.

**Comment by zoontek** [comment]
I have read the CLA Document and I hereby sign the CLA

File: android/app/build.gradle (hunk 1 above)
Comment (zoontek) [inline android/app/build.gradle:212]: Needed to extends the `Theme.SplashScreen` theme in `styles.xml`

File: android/app/src/main/AndroidManifest.xml (hunk 2 above)
Comment (zoontek) [inline android/app/src/main/AndroidManifest.xml:36]: Remove the need for a custom Activity since `react-native-bootsplash` v4 uses Android `SplashScreen` API.

File: android/app/src/main/res/values/colors.xml (hunk 3 above)
Comment (zoontek) [inline android/app/src/main/res/values/colors.xml:3]: Fix status bar being a different color than the splash screen background.

File: android/app/src/main/res/values/styles.xml (hunk 4 above)
Comment (zoontek) [inline android/app/src/main/res/values/styles.xml:10]: `android:windowLightStatusBar` and `android:windowLightNavigationBar` are not available on Android 21 (the min version), `tools:ignore` silents the Android studio linter.

File: src/App.js [hunk 11]
```diff
@@ -22,20 +22,21 @@ LogBox.ignoreLogs([
...
-    <ComposeProviders
-        components={[
-            OnyxProvider,
-            SafeAreaProvider,
-            SafeArea,
-            LocaleContextProvider,
-            HTMLEngineProvider,
-        ]}
-    >
+    <ErrorBoundary errorMessage="NewExpensify crash caught by error boundary">
```
Comment (zoontek) [inline src/App.js:25]: As an error could happen on context provider initialization, `ErrorBoundary` should wrap them to hide the splash screen is something goes wrong.

File: src/Expensify.js [hunk 12]
```diff
@@ -81,12 +82,8 @@ class Expensify extends PureComponent {
...
-                    this.hideSplash();
```
Comment (zoontek) [inline src/Expensify.js:87]: We wait for `react-navigation` to be ready. So there's no need to hide the splash screen here.

File: src/components/ErrorBoundary/BaseErrorBoundary.js [hunk 13]
```diff
@@ -2,18 +2,18 @@ import React from 'react';
...
+    /* A message posted to `onError` (along with error data) when this component intercepts an error */
```
Comment (zoontek) [inline src/components/ErrorBoundary/BaseErrorBoundary.js:5]: Renamed because it's not only used for logging anymore.
Naming taken from https://github.com/bvaughn/react-error-boundary

File: src/components/ErrorBoundary/index.native.js (hunk 10 above)
Comment (zoontek) [inline src/components/ErrorBoundary/index.native.js:17]: Hide the splash screen in case something wrong happened (ideally, the `ErrorBoundary` should show an error screen with a CTA button)

**Comment by zoontek** [comment]
Even if it wasn't implemented in this PR, I strongly recommend adding a default error screen (even without a CTA, just asking the user to restart the app) on error boundary catched error. Previously, if it happen during init (Onyx, etc) user might be stuck on the splash screen without any explanation. Now the splash screen will always disappear (at least, if it's not a complete native thread crash, just a JS error) to having something than a blank screen could be nicer 🙂

**Comment by zoontek** [comment]
Added screenshots for web / desktop where the splash screen module is a noop.

**Comment by roryabraham** [comment]
@zoontek You've got a merge conflict

**Comment by zoontek** [comment]
@roryabraham Fixed

**Comment by roryabraham** [comment]
I think we need to tag in @Expensify/design for a review here since you changed the `appBG` color.

**Comment by shawnborton** [comment]
It looks like the app background is now gray instead of white. The background of the right chat area should be white.

**Comment by roryabraham** [comment]
> Even if it wasn't implemented in this PR, I strongly recommend adding a default error screen (even without a CTA, just asking the user to restart the app) on error boundary catched error.

Good suggestion, I've created an [issue](https://github.com/Expensify/App/issues/65520) to implement this suggestion.

**Comment by zoontek** [comment]
> It looks like the app background is now gray instead of white. The background of the right chat area should be white.

Does the sign in process should also have a white background then? (It was grey, I reused this color for all screens, I can switch back to appBG: white for all screens, or just keep the subit change (sign in in grey -> loading view in white -> authenticated app in white), but at the colors are too close it felt a but weird

**Comment by shawnborton** [comment]
Let's keep all of the colors as they were before this PR was created. Thanks!

**Comment by zoontek** [comment]
> Let's keep all of the colors as they were before this PR was created. Thanks!

No problem, I will made the change tomorrow.

**Comment by zoontek** [comment]
@shawnborton @roryabraham Colors are rolled back, I fixed the `Podfile.lock` conflict, updated the screenshots

**Comment by shawnborton** [comment]
Thank you!

**Review by shawnborton — DISMISSED** [review]

**Comment by mallenexpensify** [comment]
@zoontek I think you should be compensated for[ reporting this](https://github.com/Expensify/App/pull/6478#issuecomment-979969778) above so I added a [comment to this issue.](https://github.com/Expensify/App/issues/6552#issuecomment-985932053). Payment will be issues via Upwork once the the issue is fixed internally or by a contributor

**Comment by roryabraham** [comment]
@zoontek You've got a merge conflict here

File: src/App.js [hunk 14]
```diff
@@ -25,16 +25,22 @@ LogBox.ignoreLogs([
...
             SafeArea,
             LocaleContextProvider,
-            HTMLEngineProvider,
         ]}
     >
         <CustomStatusBar />
+
         <ErrorBoundary errorMessage="NewExpensify crash caught by error boundary">
-            <Expensify />
+            <ComposeProviders
```
Comment (zoontek) [inline src/App.js:36]: As an error could happen on context provider initialization, `ErrorBoundary` should wrap them to hide the splash screen is something goes wrong.

Comment (kidroca) [reply src/App.js:36]: If there's something that might cause an Error in these components can you point it out? 

The parameters used in these components are static, so if an error is to happen it would happen all the time. Meaning if someone introduces and error they'll see it right away

If this change is accepted we should at least add a reminder doc in `ErrorBoundary` that certain context based functionality is not available inside

Though I would prefer to keep the error boundary access the full context (e.g. `withNetwork`, `withPersonalDetails` and others), and only relinquish it if there's a proven possibility that certain context can cause an error

File: src/components/ErrorBoundary/index.native.js (hunk 10 above)
Comment (kidroca) [reply src/components/ErrorBoundary/index.native.js:17]: I think this should be a part of the `BaseErrorBoundary` 
We want to hide the splash here no matter the platform

File: src/libs/BootSplash/index.native.js [hunk 15]
```diff
@@ -1,3 +1,15 @@
...
-export default RNBootSplash;
+export default {
+    // react-native-bootsplash 3.x could reject if not init (ex: when opening app via iOS >= 15 notifications)
+    // As the splash screen is not visible on these cases, we chose to ignore these errors.
+    hide: config => RNBootSplash.hide(config).catch((error) => {
+        Log.alert('[BootSplash] hiding failed', {message: error.message, error}, false);
+    }),
+    show: config => RNBootSplash.show(config).catch((error) => {
+        Log.alert('[BootSplash] showing failed', {message: error.message, error}, false);
+    }),
```
Comment (kidroca) [inline src/libs/BootSplash/index.native.js:12]: If we're no longer using `show`, and won't be able to use it in the next major version we might as well remove it

This would leave only a single `Log` call here, so we might leave it in the original place where Log is imported anyway

Also this `export` configuration does not match the style guide. I'm think it's specified somewhere, but we first define `functions` and then export them

File: src/Expensify.js [hunk 16]
```diff
@@ -102,28 +100,41 @@ class Expensify extends PureComponent {
...
+            if (state.isNavigationReady) {
+                return state;
+            }
+
+            return {
+                ...state,
+                isNavigationReady: true,
+            };
+        });
+    }
```
Comment (kidroca) [inline src/Expensify.js:133]: This is the same as

```js
setNavigationReady() {
  this.setState({ isNavigationReady: true });
}
```

When the state is already `true` it would just do nothing

But do we expect multiple `onReady` calls to be triggered anyway?

**Review by kidroca — CHANGES_REQUESTED** [review]
Some items that I've noticed 

I think addressing them would further reduce the changes in this PR

File: src/App.js (hunk 14 above)
Comment (zoontek) [reply src/App.js:36]: `react-native-onyx` might have a bug, no software is perfect. To display an error fallback, it's not worth it to try (imho)

File: src/Expensify.js (hunk 16 above)
Comment (zoontek) [reply src/Expensify.js]: Didn't see that it was a `PureComponent`, so it performs shallow compare on states comparaison anyway indeed. And lie you say, it shouldn't be called more than once.

**Comment by zoontek** [comment]
@kidroca I've taken a few feedbacks, wrote some comments for the rest.

File: src/App.js (hunk 14 above)
Comment (kidroca) [reply src/App.js:36]: Yeah, the code in `ErrorBoundary` might cause an error as well, we're obviously not adding another `ErrorBoundary` to guard from that

In the future more context components will be created it would be easier if those are kept on the same place and on the same level 
- people won't wonder where to put them 
- people would not mistakenly put them to the incorrect place

In the feature more components will be created and some of these component would rely on context, 
Adding such a component (Like a network status indicator) in the error boundary would not work for seemingly no reason

If there's a bug in the initialization of some Context it fail with a bang any way and never make it to production. Wrapping the error boundary around the context is not necessary at least the currently defined contexts

---

To sum it up Moving the `ErrorBoundary` up would work, but introduces restrictions, while it's uncertain that it will help
Not changing the Error boundary position would work as well, and it's not necessary to move it to fix the splash screen bug

File: src/Expensify.js [hunk 17]
```diff
@@ -102,28 +100,32 @@ class Expensify extends PureComponent {
...
-            this.showSplash();
-        }
+        if (this.state.isNavigationReady && this.state.isSplashShown) {
+            const authStackReady = this.props.initialReportDataLoaded && this.props.isSidebarLoaded;
+            const shouldHideSplash = !this.isAuthenticated() || authStackReady;
 
-        if (this.getAuthToken() && this.props.initialReportDataLoaded && this.props.isSidebarLoaded) {
-            this.hideSplash();
+            if (shouldHideSplash) {
+                this.hideSplash();
```
Comment (NikkiWines) [inline src/Expensify.js:113]: Since `hideSplash()` is only called once I don't think it needs to be its own separate function anymore.

File: src/App.js (hunk 14 above)
Comment (NikkiWines) [reply src/App.js:36]: I agree with @kidroca here, this change is accounting for an error that we haven't even encountered yet and it has larger implications than just catching potential initialization errors.

**Review by NikkiWines — COMMENTED** [review]
Some small comments but overall looking pretty good

**Comment by zoontek** [comment]
@NikkiWines I've taken your feedbacks 🙂

File: src/Expensify.js [hunk 18]
```diff
@@ -102,28 +99,34 @@ class Expensify extends PureComponent {
...
         AppState.removeEventListener('change', this.initializeClient);
     }
 
-    getAuthToken() {
-        return lodashGet(this.props, 'session.authToken', null);
+    setNavigationReady() {
+        this.setState({isNavigationReady: true});
+    }
+
+    isAuthenticated() {
```
Comment (roryabraham) [inline src/Expensify.js:127]: Should have a JSdoc for the return type

**Review by roryabraham — DISMISSED** [review]

File: src/Expensify.js (hunk 18 above)
Comment (zoontek) [reply src/Expensify.js:130]: Added ✅

**Review by NikkiWines — DISMISSED** [review]
👍  Looks good

**Review by roryabraham — DISMISSED** [review]

**Comment by NikkiWines** [comment]
@zoontek this PR looks good but, unfortunately, the commits aren't signed so we're unable to merge this. Would you mind re-implementing the same changes in a new PR with signed commits? Sorry about that.

**Comment by roryabraham** [comment]
@zoontek I think it might also be possible to retroactively sign your commits via rebase: https://webdevstudios.com/2020/05/26/retroactively-sign-git-commits/

**Comment by zoontek** [comment]
@NikkiWines @roryabraham Thanks for the article! I signed every commit. The reviews has been dismissed, it will need your approval again (sorry about that)

**Comment by NikkiWines** [comment]
@zoontek looks like those commits are unverified

**Comment by zoontek** [comment]
> @zoontek looks like those commits are unverified

This is weird, they were two days ago 🤔. Maybe because I pushed unverified commits on another repos since. I will re-perform the operation tomorrow.

**Comment by kidroca** [comment]
Maybe you changed your key or perhaps enabled some strict settings that makes it expire (or created a key that would expire?)
image

**Comment by zoontek** [comment]
> Maybe you changed your key or perhaps enabled some strict settings that makes it expire (or created a key that would expire?)
> image

Ah yes, I disabled it after, I didn't know it will do that.

**Comment by zoontek** [comment]
It's all good now, sorry about the inconvenience. It was my first time using GPG with git.

**Review by roryabraham — APPROVED** [review]
